src/java_bytecode/java_types.cpp

```diff
@@ -241,6 +291,83 @@ typet java_type_from_string(const std::string &src)
         return nil_typet();
       std::string class_name=src.substr(1, src.size()-2);
 
+      std::size_t f_pos=src.find('<', 1);
```

Comment:
This looks very error-prone. Is there no better way of parsing this? At least some unit tests would be good.

Reply:
why would it be error prone? The idea is to identify corresponding opening and closing `>` to see where the type variables are defined or substituted.
The `java_generics_find_closing` deals this by counting the opening and closing `>`.